File: expo-tucus/services/yandex.js
const BASE = 'https://api.music.yandex.net';

const headers = (token) => ({
  Authorization: `OAuth ${token}`,
  'User-Agent': 'Yandex-Music-API/2.0',
  'X-Yandex-Music-Client': 'YandexMusicAPI/2.0',
  Accept: 'application/json',
  'Accept-Language': 'ru,en',
});

const parseTrack = (t) => {
  const id = t.id != null ? String(t.id) : null;
  if (!id) return null;
  const artists = (t.artists || []).map(a => ({
    id: a.id != null ? String(a.id) : '',
    name: a.name || 'Unknown',
  }));
  const artist = artists.map(a => a.name).join(', ') || 'Unknown Artist';
  const coverUri = t.coverUri || '';
  const thumbnail = coverUri ? `https://${coverUri.replace('%%', '400x400')}` : '';
  return {
    id, title: t.title || 'Unknown', artist,
    artistId: artists[0]?.id || null,
    artists: artists.length ? artists : null,
    duration: (t.durationMs || 0) / 1000,
    thumbnail, source: 'yandex',
  };
};

const parseAlbum = (a) => {
  const id = a.id != null ? String(a.id) : null;
  if (!id) return null;
  const coverUri = a.coverUri || '';
  const thumbnail = coverUri ? `https://${coverUri.replace('%%', '400x400')}` : '';
  const artists = (a.artists || []).map(ar => ar.name).join(', ');
  return { id, title: a.title || 'Unknown', thumbnail, artist: artists || 'Unknown', trackCount: a.trackCount || 0, year: a.year, albumType: a.type };
};

const parseArtist = (a) => {
  const id = a.id != null ? String(a.id) : null;
  if (!id) return null;
  const cover = a.cover || {};
  const coverUri = cover.uri || a.coverUri || '';
  const thumbnail = coverUri ? `https://${coverUri.replace('%%', '400x400')}` : '';
  return { id, name: a.name || 'Unknown', thumbnail, genres: (a.genres || []).map(g => g.name || g) };
};

export const yandex = {
  async search(query, token) {
    const enc = encodeURIComponent(query);
    const res = await fetch(`${BASE}/search?text=${enc}&type=track&page=0&pageSize=20`, { headers: headers(token) });
    if (!res.ok) throw new Error(`Yandex search failed: ${res.status}`);
    const json = await res.json();
    const items = json.result?.tracks?.results || json.result?.tracks?.items || [];
    return items.map(parseTrack).filter(Boolean);
  },

  async searchAll(query, token) {
    const enc = encodeURIComponent(query);
    const res = await fetch(`${BASE}/search?text=${enc}&type=all&page=0&pageSize=20`, { headers: headers(token) });
    if (!res.ok) throw new Error('Search failed');
    const json = await res.json();
    const r = json.result || {};
    return {
      tracks: (r.tracks?.results || r.tracks?.items || []).map(parseTrack).filter(Boolean),
      albums: (r.albums?.results || r.albums?.items || []).map(parseAlbum).filter(Boolean),
      artists: (r.artists?.results || r.artists?.items || []).map(parseArtist).filter(Boolean),
    };
  },

  async getStreamUrl(trackId, token) {
    const infoRes = await fetch(`${BASE}/tracks/${trackId}/download-info`, { headers: headers(token) });
    const infoJson = await infoRes.json();
    const arr = infoJson.result || [];
    const info = arr.find(i => i.codec === 'mp3') || arr[0];
    if (!info?.downloadInfoUrl) throw new Error('No download info');

    const xmlRes = await fetch(info.downloadInfoUrl, { headers: headers(token) });
    const xml = await xmlRes.text();
    const extract = (tag) => {
      const m = xml.match(new RegExp(`<${tag}>(.*?)</${tag}>`));
      return m ? m[1] : null;
    };
    const host = extract('host'), path = extract('path'), ts = extract('ts'), s = extract('s');
    if (!host || !path || !ts || !s) throw new Error('Failed to parse XML');

    const pathClean = path.startsWith('/') ? path.slice(1) : path;
    // Simple MD5 - we'll use a JS implementation
    const hash = await md5(`XGRlBW9FXlekgbPrRHuSiA${pathClean}${s}`);
    return `https://${host}/get-mp3/${hash}/${ts}${path}`;
  },

  async getPlaylist(owner, playlistId, token) {
    const h = { ...headers(token), 'X-Yandex-Music-Client': 'YandexMusic/Android' };
    const res = await fetch(`${BASE}/users/${owner}/playlists/${playlistId}`, { headers: h });
    const json = await res.json();
    const r = json.result || {};
    const coverUri = r.cover?.uri || '';
    return {
      title: r.title || 'Playlist',
      ownerName: r.owner?.name || 'Yandex Music',
      coverUrl: coverUri ? `https://${coverUri.replace('%%', '400x400')}` : '',
      trackCount: r.trackCount || 0,
      tracks: (r.tracks || r.items || []).map(i => parseTrack(i.track)).filter(Boolean),
    };
  },

  async getLikedTracks(token) {
    const uid = await getUid(token);
    const res = await fetch(`${BASE}/users/${uid}/likes/tracks`, { headers: headers(token) });
    const json = await res.json();
    const tracks = json.result?.library?.tracks || [];
    const result = [];
    for (const item of tracks.slice(0, 30)) {
      if (item.id) {
        const tRes = await fetch(`${BASE}/tracks/${item.id}`, { headers: headers(token) });
        const tJson = await tRes.json();
        const t = parseTrack(tJson.result?.[0]);
        if (t) result.push(t);
      }
    }
    return result;
  },

  async getMyWave(token) {
    const res = await fetch(`${BASE}/rotor/station/user:onyourwave/tracks`, { headers: headers(token) });
    const json = await res.json();
    return (json.result?.sequence || []).map(i => parseTrack(i.track)).filter(Boolean);
  },

  async getSimilar(trackId, token) {
    const res = await fetch(`${BASE}/tracks/${trackId}/similar`, { headers: headers(token) });
    const json = await res.json();
    return (json.result?.similarTracks || []).map(parseTrack).filter(Boolean);
  },

  async likeTrack(trackId, token, remove) {
    const uid = await getUid(token);
    const body = JSON.stringify({ 'track-ids': [trackId] });
    const url = remove ? `${BASE}/users/${uid}/likes/tracks/remove` : `${BASE}/users/${uid}/likes/tracks/add-multiple`;
    await fetch(url, { method: 'POST', headers: { ...headers(token), 'Content-Type': 'application/json' }, body });
  },

  async getLyrics(trackId, token) {
    const urls = [
      `${BASE}/tracks/${trackId}/lyrics?format=LRC&timeStamped=true`,
      `${BASE}/tracks/${trackId}/supplement`,
    ];
    for (const url of urls) {
      const res = await fetch(url, { headers: headers(token) });
      if (!res.ok) continue;
      const json = await res.json();
      const downloadUrl = json.result?.lyrics?.downloadUrl || json.result?.downloadUrl;
      if (downloadUrl) {
        const lrcRes = await fetch(downloadUrl);
        const lrcText = await lrcRes.text();
        const lines = parseLRC(lrcText);
        if (lines.length) return lines;
      }
      const fullLyrics = json.result?.lyrics?.fullLyrics || json.result?.fullLyrics || json.result?.lyrics?.text;
      if (fullLyrics) {
        const lines = parseLRC(fullLyrics);
        if (lines.length) return lines;
        return fullLyrics.split('\n').filter(l => l.trim()).map(l => ({ text: l, time: null, duration: null }));
      }
    }
    return [];
  },
};

async function getUid(token) {
  const res = await fetch(`${BASE}/account/status`, { headers: headers(token) });
  const json = await res.json();
  return String(json.result?.account?.uid || '');
}

// Simple MD5 implementation
async function md5(string) {
  const msgUint8 = new TextEncoder().encode(string);
  const hashBuffer = await crypto.subtle.digest('MD5', msgUint8).catch(() => null);
  if (hashBuffer) {
    return Array.from(new Uint8Array(hashBuffer)).map(b => b.toString(16).padStart(2, '0')).join('');
  }
  // Fallback: simple hash
  let hash = 0;
  for (let i = 0; i < string.length; i++) {
    const char = string.charCodeAt(i);
    hash = ((hash << 5) - hash) + char;
    hash |= 0;
  }
  return Math.abs(hash).toString(16).padStart(32, '0');
}

function parseLRC(text) {
  const lines = [];
  const lineTimeRe = /\[(\d+):(\d+)[.:](\d+)\]/g;
  const plainTimeRe = /\[(\d+):(\d+)\]/;

  for (const raw of text.split('\n')) {
    const line = raw.trim();
    if (!line) continue;

    const times = [];
    let m;
    const tempLine = line;
    while ((m = lineTimeRe.exec(tempLine)) !== null) {
      const min = parseInt(m[1]), sec = parseInt(m[2]);
      const csStr = m[3];
      let ms = parseInt(csStr);
      if (csStr.length === 1) ms *= 100;
      else if (csStr.length === 2) ms *= 10;
      times.push(min * 60 + sec + ms / 1000);
    }
    lineTimeRe.lastIndex = 0;

    if (times.length === 0) {
      const pm = line.match(plainTimeRe);
      if (pm) {
        const content = line.slice(pm[0].length).trim();
        if (content) lines.push({ text: content, time: parseInt(pm[1]) * 60 + parseInt(pm[2]), duration: null });
      } else if (!line.startsWith('[')) {
        lines.push({ text: line, time: null, duration: null });
      }
      continue;
    }

    let content = line.replace(/\[\d+:\d+[.:]\d+\]/g, '').trim();
    for (const t of times) {
      lines.push({ text: content, time: t, duration: null });
    }
  }

  // Calculate durations
  for (let i = 0; i < lines.length; i++) {
    if (lines[i].time != null) {
      const next = lines[i + 1]?.time ?? (lines[i].time + 5);
      lines[i].duration = next - lines[i].time;
    }
  }

  return lines;
}

File: src-tauri/src/yandex.rs
use crate::{Track, LyricLine};
use serde_json::Value;

const BASE: &str = "https://api.music.yandex.net";
const TIMEOUT: u64 = 10;

fn headers(token: &str) -> Vec<(String, String)> {
    vec![
        ("Authorization".into(), format!("OAuth {}", token)),
        ("User-Agent".into(), "Yandex-Music-API/2.0".into()),
        ("Accept".into(), "application/json".into()),
    ]
}

fn parse_track(t: &Value) -> Option<Track> {
    let id = t.get("id")?.as_i64()?.to_string();
    let title = t.get("title")?.as_str()?.to_string();
    let artists = t.get("artists")?.as_array()?;
    let artist: String = artists
        .iter()
        .filter_map(|a| a.get("name")?.as_str())
        .collect::<Vec<_>>()
        .join(", ");
    let cover = t.get("coverUri")?.as_str()?;
    let thumbnail = format!("https://{}", cover.replace("%%", "400x400"));
    let duration_ms = t.get("durationMs")?.as_f64()?;
    Some(Track {
        id,
        title,
        artist,
        source: "yandex".into(),
        thumbnail: Some(thumbnail),
        duration: Some(duration_ms / 1000.0),
    })
}

pub async fn search(query: &str, token: &str) -> Result<Vec<Track>, String> {
    let client = reqwest::Client::new();
    let url = format!(
        "{}/search?text={}&type=track&page=0&pageSize=20",
        BASE,
        urlencoding::encode(query)
    );
    let resp = client
        .get(&url)
        .headers(
            headers(token)
                .into_iter()
                .map(|(k, v)| (k.parse().unwrap(), v.parse().unwrap()))
                .collect(),
        )
        .timeout(std::time::Duration::from_secs(TIMEOUT))
        .send()
        .await
        .map_err(|e| e.to_string())?;
    let json: Value = resp.json().await.map_err(|e| e.to_string())?;
    let results = json
        .pointer("/result/tracks/results")
        .or_else(|| json.pointer("/result/tracks/items"))
        .and_then(|v| v.as_array())
        .cloned()
        .unwrap_or_default();
    Ok(results.iter().filter_map(parse_track).collect())
}

pub async fn get_stream_url(track_id: &str, token: &str) -> Result<String, String> {
    let client = reqwest::Client::new();
    let url = format!("{}/tracks/{}/download-info", BASE, track_id);
    let resp = client
        .get(&url)
        .headers(
            headers(token)
                .into_iter()
                .map(|(k, v)| (k.parse().unwrap(), v.parse().unwrap()))
                .collect(),
        )
        .timeout(std::time::Duration::from_secs(TIMEOUT))
        .send()
        .await
        .map_err(|e| e.to_string())?;
    let json: Value = resp.json().await.map_err(|e| e.to_string())?;
    let arr = json["result"].as_array().ok_or("No download info")?;
    let info = arr
        .iter()
        .find(|i| i["codec"].as_str() == Some("mp3"))
        .or_else(|| arr.first())
        .ok_or("No suitable codec")?;
    let download_url = info["downloadInfoUrl"]
        .as_str()
        .ok_or("No download URL")?;

    let xml_resp = client
        .get(download_url)
        .headers(
            headers(token)
                .into_iter()
                .map(|(k, v)| (k.parse().unwrap(), v.parse().unwrap()))
                .collect(),
        )
        .timeout(std::time::Duration::from_secs(TIMEOUT))
        .send()
        .await
        .map_err(|e| e.to_string())?;
    let xml = xml_resp.text().await.map_err(|e| e.to_string())?;

    let extract = |tag: &str| -> Option<String> {
        let pattern = format!("<{}>(.*?)</{}>", tag, tag);
        let re = regex::Regex::new(&pattern).ok()?;
        let caps = re.captures(&xml)?;
        caps.get(1).map(|m| m.as_str().to_string())
    };

    let host = extract("host").ok_or("No host")?;
    let path = extract("path").ok_or("No path")?;
    let ts = extract("ts").ok_or("No ts")?;
    let s = extract("s").ok_or("No s")?;

    let path_clean = path.strip_prefix('/').unwrap_or(&path);
    let hash_input = format!("XGRlBW9FXlekgbPrRHuSiA{}{}", path_clean, s);
    let hash = format!("{:x}", md5::compute(hash_input));

    Ok(format!("https://{}/get-mp3/{}/{}{}", host, hash, ts, path))
}

pub async fn get_lyrics(track_id: &str, token: &str) -> Result<Vec<LyricLine>, String> {
    let client = reqwest::Client::new();
    let urls = vec![
        format!("{}/tracks/{}/lyrics?format=LRC&timeStamped=true", BASE, track_id),
        format!("{}/tracks/{}/supplement", BASE, track_id),
    ];

    for url in urls {
        if let Ok(resp) = client
            .get(&url)
            .headers(
                headers(token)
                    .into_iter()
                    .map(|(k, v)| (k.parse().unwrap(), v.parse().unwrap()))
                    .collect(),
            )
            .timeout(std::time::Duration::from_secs(TIMEOUT))
            .send()
            .await
        {
            if let Ok(json) = resp.json::<Value>().await {
                let download_url = json["result"]["lyrics"]["downloadUrl"]
                    .as_str()
                    .or_else(|| json["result"]["downloadUrl"].as_str());
                if let Some(durl) = download_url {
                    if let Ok(lrc_resp) = client
                        .get(durl)
                        .timeout(std::time::Duration::from_secs(TIMEOUT))
                        .send()
                        .await
                    {
                        if let Ok(lrc_text) = lrc_resp.text().await {
                            let lines = parse_lrc(&lrc_text);
                            if !lines.is_empty() {
                                return Ok(lines);
                            }
                        }
                    }
                }
                if let Some(full) = json["result"]["lyrics"]["fullLyrics"]
                    .as_str()
                    .or_else(|| json["result"]["fullLyrics"].as_str())
                    .or_else(|| json["result"]["lyrics"]["text"].as_str())
                {
                    let lines = parse_lrc(full);
                    if !lines.is_empty() {
                        return Ok(lines);
                    }
                    return Ok(full
                        .lines()
                        .filter(|l| !l.trim().is_empty())
                        .map(|l| LyricLine {
                            text: l.to_string(),
                            time: None,
                            duration: None,
                        })
                        .collect());
                }
            }
        }
    }
    Ok(vec![])
}

fn parse_lrc(text: &str) -> Vec<LyricLine> {
    let mut lines = Vec::new();
    let re = regex::Regex::new(r"\[(\d+):(\d+)[.:](\d+)\]").ok();
    let plain_re = regex::Regex::new(r"\[(\d+):(\d+)\]").ok();

    for raw in text.lines() {
        let line = raw.trim();
        if line.is_empty() {
            continue;
        }

        let mut times = Vec::new();
        if let Some(re) = &re {
            for cap in re.captures_iter(line) {
                let min: f64 = cap[1].parse().unwrap_or(0.0);
                let sec: f64 = cap[2].parse().unwrap_or(0.0);
                let cs_str = &cap[3];
                let ms: f64 = cs_str.parse().unwrap_or(0.0);
                let ms = if cs_str.len() == 1 {
                    ms * 100.0
                } else if cs_str.len() == 2 {
                    ms * 10.0
                } else {
                    ms
                };
                times.push(min * 60.0 + sec + ms / 1000.0);
            }
        }

        if times.is_empty() {
            if let Some(pm) = plain_re.as_ref().and_then(|r| r.captures(line)) {
                let min: f64 = pm[1].parse().unwrap_or(0.0);
                let sec: f64 = pm[2].parse().unwrap_or(0.0);
                let content = line[pm[0].len()..].trim().to_string();
                if !content.is_empty() {
                    lines.push(LyricLine {
                        text: content,
                        time: Some(min * 60.0 + sec),
                        duration: None,
                    });
                }
            } else if !line.starts_with('[') {
                lines.push(LyricLine {
                    text: line.to_string(),
                    time: None,
                    duration: None,
                });
            }
            continue;
        }

        let content = line
            .replace(|c: char| c == '[' || c == ']', "")
            .split(|c: char| c.is_ascii_digit() || c == ':' || c == '.' || c == ',')
            .last()
            .unwrap_or("")
            .trim()
            .to_string();
        let content = if content.is_empty() {
            // Extract text after all timestamps
            let re2 = regex::Regex::new(r"\[\d+:\d+[.:]\d+\]").ok();
            if let Some(re2) = re2 {
                re2.replace_all(line, "").trim().to_string()
            } else {
                line.to_string()
            }
        } else {
            content
        };

        for t in times {
            lines.push(LyricLine {
                text: content.clone(),
                time: Some(t),
                duration: None,
            });
        }
    }

    // Calculate durations
    for i in 0..lines.len() {
        if let Some(t) = lines[i].time {
            let next = lines.get(i + 1).and_then(|l| l.time).unwrap_or(t + 5.0);
            lines[i].duration = Some(next - t);
        }
    }

    lines
}
